File: src/main/java/legacy/Percent.java
package legacy;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;

public final class Percent {

    private final static BigDecimal FRACTION_DIVISOR = BigDecimal.valueOf(100);

    private final BigDecimal value;

    public Percent(BigDecimal value) {
        this.value = value;
    }

    public BigDecimal getValue() {
        return value;
    }

    public BigDecimal getFractionalValue() {
        return value.divide(FRACTION_DIVISOR, this.value.scale() + 2, RoundingMode.HALF_UP);
    }

    public static Percent of(BigDecimal value) {
        return new Percent(value);
    }

    @Override
    public String toString() {
        return "legacy.Percent{" +
                "value=" + value +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Percent percent = (Percent) o;
        return Objects.equals(value, percent.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }
}

File: src/main/kotlin/refactored/Percent.kt
package refactored

import java.math.BigDecimal
import java.math.RoundingMode
import java.util.*

class Percent(val value: BigDecimal) {

    val fractionalValue: BigDecimal
        get() = value.divide(FRACTION_DIVISOR, value.scale() + 2, RoundingMode.HALF_UP)

    override fun toString(): String {
        return "legacy.Percent{" +
                "value=" + value +
                '}'
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val percent = o as Percent
        return value == percent.value
    }

    override fun hashCode(): Int {
        return Objects.hash(value)
    }

    companion object {
        private val FRACTION_DIVISOR = BigDecimal.valueOf(100)
        fun of(value: BigDecimal): Percent {
            return Percent(value)
        }
    }
}